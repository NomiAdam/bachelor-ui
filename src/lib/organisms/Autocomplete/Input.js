import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { lightTheme } from '../../constants/theme';

const StyledInput = styled.input`\n    border: none;\n    border-bottom: 1px solid ${ lightTheme.lightGrey };\n    box-sizing: border-box;\n    width: 100%;\n	min-height: 30px;\n	padding: 10px 20px;\n	font-size: 16px;\n    &:focus {\n        outline: none;\n    }\n`;

const Input = ({
    value, handleFocus, handleChange, placeholder, handleEnter,
}) => (
    <StyledInput onKeyPress={ handleEnter } onChange={ handleChange } onFocus={ handleFocus } value={ value } placeholder={ placeholder } />
);

Input.propTypes = {
    /**
     * Current value of Input field
     */
    value: PropTypes.string,
    /**
     * Function triggered when Input is Focused, ie. user clicked into Input
     */
    handleFocus: PropTypes.func,
    /**
     * Function to handle value change on user input
     */
    handleChange: PropTypes.func,
    /**
     * String placeholder value, for when Input is empty
     */
    placeholder: PropTypes.string,
    /**
     * Function to be called upon enter press
     */
    handleEnter: PropTypes.func,
};

export { StyledInput };
export default Input;
