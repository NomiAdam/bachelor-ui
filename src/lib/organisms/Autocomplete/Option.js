import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import {lightTheme} from '../../constants/theme';

const StyledOption = styled.div`
    width: 100%;
    border: none;
	padding: 10px 20px;
	background-color: ${({selected}) => selected ? lightTheme.lightGrey : 'initial'};
    &:hover {
        background-color: ${lightTheme.lightGrey};
        cursor: pointer;
    }
`;

const Option = ({value, onClick, selected}) => (
    <StyledOption onMouseDown={onClick(value)} selected={selected}>
        {value}
    </StyledOption>
);

Option.propTypes = {
    /**
     * String value of Option
     */
    value: PropTypes.string,
    /**
     * onMouseDown event handler, (value) => (e) =>
     */
    onClick: PropTypes.func.isRequired,
    /**
     * Boolean value whether current option is selected or not
     */
    selected: PropTypes.bool,
};

export {StyledOption};
export default Option;
