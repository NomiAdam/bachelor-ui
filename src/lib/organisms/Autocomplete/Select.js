import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import {
    map, length, equals, o,
} from 'ramda';
import Option from './Option';
import { lightTheme } from '../../constants/theme';

const StyledSelect = styled.div`
    display: ${ props => (props.visible ? 'block' : 'none') }
    position: absolute;
    z-index: 999;
    background-color: ${ lightTheme.white };
    width: 100%;
    min-height: 0px;
    max-height: 500px;
    border: 1px solid ${ lightTheme.lightGrey }; 
    border-top: none;
    border-bottom-right-radius: 4px;
    border-bottom-left-radius: 4px;
    ${ props => (props.isEmpty ? 'border: none;' : '') }
    font-weight: 300;
    font-size: 16px;
    overflow-y: auto;
`;

const doesItHaveOptions = o(equals(0), length);
const Select = ({
    visible, suggestions, onClick, selected,
}) => {
    const equalsSelected = equals(selected);
    const mapSuggestion = ({ value }) => (
        <Option value={ value } onClick={ onClick } key={ value } selected={ equalsSelected(value) } />
    );
    const mappedOptions = map(mapSuggestion);
    return (
        <StyledSelect visible={ visible } isEmpty={ doesItHaveOptions(suggestions) }>
            {visible && mappedOptions(suggestions)}
        </StyledSelect>
    );
};

Select.propTypes = {
    /**
     * Whether select div is visible or not, is visible when Input is focused
     */
    visible: PropTypes.bool,
    /**
     * Array of possible values [ { value: 'Value', label: 'Label' }, ... ]
     */
    suggestions: PropTypes.array,
    /**
     * Option onMouseDown handler
     */
    onClick: PropTypes.func,
    /**
     * String definition of selected suggestion value
     */
    selected: PropTypes.string,
};

Select.defaultProps = {
    suggestions: [],
};

export { StyledSelect };
export default Select;
