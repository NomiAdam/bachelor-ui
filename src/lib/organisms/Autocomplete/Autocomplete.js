import React, {Component} from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import {
    path,
    contains,
    filter,
    o,
    prop,
    toLower,
    head,
    propOr,
    propEq,
    when,
    findIndex,
    dec,
    add,
    nth,
} from 'ramda';
import {isNilOrEmptyString, isNotNil} from 'ramda-extension';
import Input from './Input';
import Select from './Select';

const getInputValue = path(['target', 'value']);

const getFirstSuggestionValue = o(propOr('', 'value'), head);
const getCurrentIndex = (value) => findIndex(propEq('value', value));
const getNextValue = when(isNotNil, prop('value'));

const StyledAutocomplete = styled.div`
    width: 100%;
    position: relative;
`;

class Autocomplete extends Component {
    state = {
        value: this.props.value,
        isFocused: false,
        selected: '',
        filteredSuggestions: this.props.suggestions,
    };

    componentDidMount() {
        const {value, suggestions} = this.props;
        if (isNilOrEmptyString(value)) {
            this.setState({selected: getFirstSuggestionValue(suggestions)})
        } else {
            this.handleStateChange(value);
            this.setState({selected: value})
        }
    };

    handleFocus = () => {
        this.setState({isFocused: true});
    };

    handleBlur = () => {
        this.setState({isFocused: false});
    };

    handleValueChange = (e) => this.handleStateChange(getInputValue(e));

    handleSelect = (value) => () => {
        this.handleStateChange(value);
        const {handleChange} = this.props;
        handleChange(value);
        this.handleBlur();
    };

    handleStateChange = (value) => {
        const {suggestions} = this.props;

        const matchString = contains(toLower(value));
        const validatedMatchingValue = o(toLower, prop('value'));
        const matchFilter = o(matchString, validatedMatchingValue);
        const validatedSuggestions = filter(matchFilter);

        this.setState({
            value,
            filteredSuggestions: validatedSuggestions(suggestions)
        });
    };

    handleKeyPress = (e) => {
        const {selected} = this.state;
        const {suggestions} = this.props;
        const index = getCurrentIndex(selected)(suggestions);
        switch (e.key) {
            case 'ArrowUp': {
                const getValueOfNext = nth(dec(index, 1), suggestions);
                this.setState({selected: getNextValue(getValueOfNext) || selected});
            }
                break;
            case 'ArrowDown': {
                const getValueOfNext = nth(add(index, 1), suggestions);
                this.setState({selected: getNextValue(getValueOfNext) || selected});
            }
                break;
            case 'Enter':
                this.handleSelect(selected)(e);
                break;
        }
    };

    render() {
        const {placeholder} = this.props;
        const {isFocused, value, selected, filteredSuggestions} = this.state;
        return (
            <StyledAutocomplete onBlur={this.handleBlur} onKeyDown={this.handleKeyPress}>
                <Input
                    value={value}
                    handleFocus={this.handleFocus}
                    handleChange={this.handleValueChange}
                    handleEnter={this.handleEnter}
                    placeholder={placeholder}
                />
                <Select
                    visible={isFocused}
                    suggestions={filteredSuggestions}
                    onClick={this.handleSelect}
                    selected={selected}
                    handleKeyPress={this.handleKeyPress}
                />
            </StyledAutocomplete>
        );

    }
}

Autocomplete.propTypes = {
    /**
     * Array of possible values [ { value: 'Value', label: 'Label' }, ... ]
     */
    suggestions: PropTypes.array,
    /**
     * Current value of Input component
     */
    value: PropTypes.string,
    /**
     * Function to handle Change on Input or Select
     */
    handleChange: PropTypes.func,
    /**
     * String definition of placeholder
     */
    placeholder: PropTypes.string,
};

Autocomplete.defaultProps = {
    value: '',
    suggestions: [],
    placeholder: '',
};

export {StyledAutocomplete};
export default Autocomplete;
