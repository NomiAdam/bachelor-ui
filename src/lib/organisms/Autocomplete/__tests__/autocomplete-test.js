/* eslint-disable no-undef */
import React from 'react';
import Autocomplete, {StyledAutocomplete} from '../Autocomplete';
import {shallow} from 'enzyme';

const mock = jest.fn();
const mockSuggestions = [{value: 'Value', label: 'Label'}];

describe('Autocomplete', () => {
    beforeEach(() => {
        jest.resetAllMocks();
    });
    it('should be defined', () => {
        expect(Autocomplete).toBeDefined();
    });
    it('should match the snapshot', () => {
        expect(shallow(<Autocomplete/>)).toMatchSnapshot();
    });
    it('should match the snapshot with suggestions', () => {
        expect(shallow(<Autocomplete suggestions={mockSuggestions}/>)).toMatchSnapshot();
    });
    it('should be of type StyledAutocomplete', () => {
        expect(shallow(<Autocomplete/>).type()).toBe(StyledAutocomplete);
    });
    it('should have default state', () => {
        expect(shallow(<Autocomplete />).state()).toEqual(
            {
                isFocused: false,
                value: '',
                selected: '',
                filteredSuggestions: [],
            }
        );
    });
    it('should not fire action on render', () => {
        shallow(<Autocomplete handleChange={mock}/>);
        expect(mock).not.toHaveBeenCalled();
    });
});
