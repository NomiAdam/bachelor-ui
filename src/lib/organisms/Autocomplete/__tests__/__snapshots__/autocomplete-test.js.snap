// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Autocomplete should match the snapshot 1`] = `
<Autocomplete__StyledAutocomplete
  onBlur={[Function]}
  onKeyDown={[Function]}
>
  <Input
    handleChange={[Function]}
    handleFocus={[Function]}
    placeholder=""
    value=""
  />
  <Select
    handleKeyPress={[Function]}
    onClick={[Function]}
    selected=""
    suggestions={Array []}
    visible={false}
  />
</Autocomplete__StyledAutocomplete>
`;

exports[`Autocomplete should match the snapshot with suggestions 1`] = `
<Autocomplete__StyledAutocomplete
  onBlur={[Function]}
  onKeyDown={[Function]}
>
  <Input
    handleChange={[Function]}
    handleFocus={[Function]}
    placeholder=""
    value=""
  />
  <Select
    handleKeyPress={[Function]}
    onClick={[Function]}
    selected="Value"
    suggestions={
      Array [
        Object {
          "label": "Label",
          "value": "Value",
        },
      ]
    }
    visible={false}
  />
</Autocomplete__StyledAutocomplete>
`;
